alf/algorithms/oac_algorithm.py

```diff
@@ -0,0 +1,220 @@
+                 max_log_alpha=None,
+                 target_update_tau=0.05,
+                 target_update_period=1,
+                 dqda_clipping=None,
+                 actor_optimizer=None,
+                 critic_optimizer=None,
+                 alpha_optimizer=None,
+                 debug_summaries=False,
+                 name="OacAlgorithm"):
+        """
+        Refer to SacAlgorithm for Args besides the following.
+
+        Args:
+            explore (bool): default is True for OAC algorithm, where only 
+                continuous action space is supported. When 'explore' is False, 
+                OAC is the same as SAC. 
+            explore_delta (float): parameter controlling how optimistic in shifting
+                the mean of the target policy to get the mean of the explore policy.
+            beta_ub (float): parameter for computing the upperbound of Q value:
+                :math:`Q_ub(s,a) = \mu_Q(s,a) + \beta_ub * \sigma_Q(s,a)`    
+        """
+        super().__init__(
+            observation_spec,
+            action_spec,
+            reward_spec=reward_spec,
+            actor_network_cls=actor_network_cls,
+            critic_network_cls=critic_network_cls,
+            q_network_cls=q_network_cls,
+            use_entropy_reward=use_entropy_reward,
+            use_parallel_network=use_parallel_network,
+            num_critic_replicas=num_critic_replicas,
+            env=env,
+            config=config,
+            critic_loss_ctor=critic_loss_ctor,
+            target_entropy=target_entropy,
+            prior_actor_ctor=prior_actor_ctor,
+            initial_log_alpha=initial_log_alpha,
+            max_log_alpha=max_log_alpha,
+            target_update_tau=target_update_tau,
+            target_update_period=target_update_period,
+            dqda_clipping=dqda_clipping,
+            actor_optimizer=actor_optimizer,
+            critic_optimizer=critic_optimizer,
+            alpha_optimizer=alpha_optimizer,
+            debug_summaries=debug_summaries,
+            name=name)
+
+        if explore:
+            assert self._act_type == ActionType.Continuous, (
+                "Only continuous action space is supported for explore mode.")
+        self._explore = explore
+        self._explore_delta = explore_delta
+        self._beta_ub = beta_ub
+
+    def _predict_action(self,
+                        observation,
+                        state: SacActionState,
+                        epsilon_greedy=None,
+                        eps_greedy_sampling=False,
+                        explore=False):
+        """
+        Differences between SacAlgorithm._predict_action:
+
+        1. Only continuous actions are supported.
+
+        2. Add a switch for explore mode where OAC explore policy is constructed 
+        from the target policy (actor_network) and used for action prediction.
+        """
+        new_state = SacActionState()
+        with torch.enable_grad():
```

Comment:
you only use enable_grad for line 160-171

Reply:
That have tried that in the first place, but it did not work. It seems that I have to include forwarding of self._actor_network into the "with torch.enable_grad()" to make things work.

Reply:
I have tried that in the first place and it did not work. I just confirmed that it seems that "with torch.enable_grad()" has to start before the "if explore" to make things work. Will update it.

Reply:
updated.